File: python/src/train/models/mixformer/encoding.py
"""MixFormer core modules: HeadMixing, SwiGLU FFN, QueryMixer, OutputFusion.

Paper §3.3: Each MixFormer block consists of QueryMixer + CrossAttention + OutputFusion.
Here we implement QueryMixer and OutputFusion with per-head SwiGLU FFN.
"""

from __future__ import annotations

import torch
import torch.nn as nn
import torch.nn.functional as F


def head_mixing(x: torch.Tensor) -> torch.Tensor:
    """Parameter-free cross-head information exchange.

    Paper §3.3.1 Eq. (3): Reshapes [B, N, D] → [B, N, N, D/N],
    transposes dims 1↔2, and flattens to [B, N, D].

    Each output head j receives chunk j from every input head,
    enabling cross-head communication at zero parameter cost.
    """
    B, N, D = x.shape
    return x.reshape(B, N, N, D // N).transpose(1, 2).reshape(B, N, D)


class SwiGLUFFN(nn.Module):
    """SwiGLU-activated feed-forward network.

    Paper §3.3: SwiGLU(x) = (SiLU(x · W_gate) ⊗ (x · W_up)) · W_down
    where ⊗ denotes element-wise multiplication.
    """

    def __init__(self, d: int, d_ff: int):
        super().__init__()
        self.gate = nn.Linear(d, d_ff, bias=False)
        self.up = nn.Linear(d, d_ff, bias=False)
        self.down = nn.Linear(d_ff, d, bias=False)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.down(F.silu(self.gate(x)) * self.up(x))


class QueryMixer(nn.Module):
    """Query Mixer: replaces self-attention with HeadMixing + per-head SwiGLU FFN.

    Paper §3.3.1 Eq. (3)–(4):
        P = HeadMixing(Norm(X)) + X
        Q_i = SwiGLUFFN_i(Norm(P_i)) + P_i
    """

    def __init__(self, d: int, d_ff: int, num_heads: int):
        super().__init__()
        self.norm1 = nn.LayerNorm(d)
        self.norm2 = nn.LayerNorm(d)
        self.head_ffns = nn.ModuleList([SwiGLUFFN(d, d_ff) for _ in range(num_heads)])

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        _, N, _ = x.shape

        residual = x
        x = self.norm1(x)
        x = head_mixing(x)
        x = x + residual

        residual = x
        x = self.norm2(x)
        outputs = [self.head_ffns[i](x[:, i, :]) for i in range(N)]
        x = torch.stack(outputs, dim=1)
        x = x + residual
        return x


class OutputFusion(nn.Module):
    """Output Fusion: per-head SwiGLU FFN + residual.

    Paper §3.3.3 Eq. (9):
        o_i = SwiGLUFFN_i(Norm(z_i)) + z_i

    Each head receives its own FFN after cross-attention aggregation.
    """

    def __init__(self, d: int, d_ff: int, num_heads: int):
        super().__init__()
        self.norm = nn.LayerNorm(d)
        self.head_ffns = nn.ModuleList([SwiGLUFFN(d, d_ff) for _ in range(num_heads)])

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        _, N, _ = x.shape

        residual = x
        x = self.norm(x)
        outputs = [self.head_ffns[i](x[:, i, :]) for i in range(N)]
        x = torch.stack(outputs, dim=1)
        x = x + residual
        return x


class MixFormerBlock(nn.Module):
    """One MixFormer block: QueryMixer + OutputFusion with Pre-LN.

    Paper §3.3:
        Z^{(l)} = OutputFusion(CrossAttn(QueryMixer(Z^{(l-1)})))
    """

    def __init__(self, d: int, d_ff: int, num_heads: int):
        super().__init__()
        self.query_mixer = QueryMixer(d, d_ff, num_heads)
        self.output_fusion = OutputFusion(d, d_ff, num_heads)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.query_mixer(x)
        x = self.output_fusion(x)
        return x

File: src/models/mixformer/encoding.rs
//! MixFormer core modules: HeadMixing, SwiGLU FFN, QueryMixer, OutputFusion.
//!
//! Paper §3.3: Each MixFormer block = QueryMixer + CrossAttention + OutputFusion.
//! CrossAttention depends on sequence data; QueryMixer and OutputFusion are
//! implemented here.

use candle_core::{Module, Result, Tensor};
use candle_nn::{layer_norm, linear_no_bias, LayerNorm, Linear, VarBuilder};

/// Parameter-free cross-head information exchange.
///
/// Paper §3.3.1: Reshapes [B, N, D] → [B, N, N, D/N],
/// transposes dims 1↔2, and flattens back to [B, N, D].
fn head_mixing(x: &Tensor, n: usize) -> Result<Tensor> {
    let (b, _n, d) = x.dims3()?;
    let chunk = d / n;
    let x = x.reshape((b, n, n, chunk))?;
    let x = x.permute((0, 2, 1, 3))?;
    x.reshape((b, n, d))
}

/// SwiGLU-activated FFN: SwiGLU(x) = (SiLU(x·W_gate)⊗(x·W_up))·W_down
struct SwiGLUFFN {
    gate: Linear,
    up: Linear,
    down: Linear,
}

impl SwiGLUFFN {
    /// Create SwiGLU FFN with gate, up, down projections.
    fn new(vb: VarBuilder, d: usize, d_ff: usize) -> Result<Self> {
        let gate = linear_no_bias(d, d_ff, vb.pp("gate"))?;
        let up = linear_no_bias(d, d_ff, vb.pp("up"))?;
        let down = linear_no_bias(d_ff, d, vb.pp("down"))?;
        Ok(Self { gate, up, down })
    }

    /// Forward: SwiGLU(x) = (SiLU(x·W_gate)⊗(x·W_up))·W_down.
    fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let gate = self.gate.forward(x)?;
        let up = self.up.forward(x)?;
        let silu = candle_nn::ops::sigmoid(&gate)?.mul(&gate)?;
        self.down.forward(&silu.mul(&up)?)
    }
}

/// Query Mixer: HeadMixing + per-head SwiGLU FFN.
///
/// Paper §3.3.1 Eq. (3)–(4):
///   P = HeadMixing(Norm(X)) + X
///   Q_i = SwiGLUFFN_i(Norm(P_i)) + P_i
pub struct QueryMixer {
    norm1: LayerNorm,
    norm2: LayerNorm,
    head_ffns: Vec<SwiGLUFFN>,
    num_heads: usize,
}

impl QueryMixer {
    /// Create a new Query Mixer.
    pub fn new(vb: VarBuilder, d: usize, d_ff: usize, num_heads: usize) -> Result<Self> {
        let norm1 = layer_norm(d, 1e-5, vb.pp("norm1"))?;
        let norm2 = layer_norm(d, 1e-5, vb.pp("norm2"))?;
        let mut head_ffns = Vec::with_capacity(num_heads);
        for i in 0..num_heads {
            head_ffns.push(SwiGLUFFN::new(vb.pp(format!("head_ffns.{}", i)), d, d_ff)?);
        }
        Ok(Self { norm1, norm2, head_ffns, num_heads })
    }

    /// Forward: HeadMixing → per-head SwiGLU FFN.
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        // HeadMixing block
        let residual = x;
        let x = self.norm1.forward(x)?;
        let x = head_mixing(&x, self.num_heads)?;
        let x = (residual + &x)?;

        // Per-head FFN
        let residual = &x;
        let x = self.norm2.forward(&x)?;
        let mut outputs = Vec::with_capacity(self.num_heads);
        for i in 0..self.num_heads {
            let head = x.narrow(1, i, 1)?.squeeze(1)?;
            outputs.push(self.head_ffns[i].forward(&head)?.unsqueeze(1)?);
        }
        let x = Tensor::cat(&outputs, 1)?;
        residual + &x
    }
}

/// Output Fusion: per-head SwiGLU FFN + residual.
///
/// Paper §3.3.3 Eq. (9):
///   o_i = SwiGLUFFN_i(Norm(z_i)) + z_i
pub struct OutputFusion {
    norm: LayerNorm,
    head_ffns: Vec<SwiGLUFFN>,
    num_heads: usize,
}

impl OutputFusion {
    /// Create a new Output Fusion layer.
    pub fn new(vb: VarBuilder, d: usize, d_ff: usize, num_heads: usize) -> Result<Self> {
        let norm = layer_norm(d, 1e-5, vb.pp("norm"))?;
        let mut head_ffns = Vec::with_capacity(num_heads);
        for i in 0..num_heads {
            head_ffns.push(SwiGLUFFN::new(vb.pp(format!("head_ffns.{}", i)), d, d_ff)?);
        }
        Ok(Self { norm, head_ffns, num_heads })
    }

    /// Forward pass.
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let residual = x;
        let x = self.norm.forward(x)?;
        let mut outputs = Vec::with_capacity(self.num_heads);
        for i in 0..self.num_heads {
            let head = x.narrow(1, i, 1)?.squeeze(1)?;
            outputs.push(self.head_ffns[i].forward(&head)?.unsqueeze(1)?);
        }
        let x = Tensor::cat(&outputs, 1)?;
        residual + &x
    }
}

/// One MixFormer block: QueryMixer → OutputFusion.
pub struct MixFormerBlock {
    query_mixer: QueryMixer,
    output_fusion: OutputFusion,
}

impl MixFormerBlock {
    /// Create a new MixFormer block.
    pub fn new(vb: VarBuilder, d: usize, d_ff: usize, num_heads: usize) -> Result<Self> {
        let query_mixer = QueryMixer::new(vb.pp("query_mixer"), d, d_ff, num_heads)?;
        let output_fusion = OutputFusion::new(vb.pp("output_fusion"), d, d_ff, num_heads)?;
        Ok(Self { query_mixer, output_fusion })
    }

    /// Forward pass: QueryMixer → OutputFusion.
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let x = self.query_mixer.forward(x)?;
        self.output_fusion.forward(&x)
    }
}
